[Core] Use platform-agnostic device control for DP engine core

jianzs opened the pull request:
Refactor the DP engine core to use the platform-specific `device_control_env_var` attribute instead of hardcoding CUDA_VISIBLE_DEVICES. This change improves platform compatibility and code maintainability.

The update includes:
1. Moving the `device_id_to_physical_device_id` function to a shared utils file
2. Updating imports in CUDA and ROCm platform files
3. Replacing CUDA-specific environment variable setting with a platform-agnostic approach in the DPEngineCoreProc class

This refactoring enhances the flexibility of the codebase to support different platforms more seamlessly.

jianzs commented:
@njhill  Hi, could you help review and move this PR forward?

njhill commented on vllm/v1/engine/core.py:
```diff
@@ -592,13 +592,13 @@ def __init__(
         assert 0 <= local_dp_rank <= dp_rank < dp_size
 
         from vllm.platforms import current_platform
```
I think we should have here
```python
    if device_control_env_var := getattr(current_platform, "device_control_env_var", None):
```

jianzs replied:
> I think we should have here
> 
> ```python
>     if device_control_env_var := getattr(current_platform, "device_control_env_var", None):
> ```

Is this check necessary given that the `device_control_env_var` attribute is already defined in the `Platform` base class?

https://github.com/vllm-project/vllm/blob/950b71186f9696a60f1a53cd1a033f7fef952500/vllm/platforms/interface.py#L92-L111

njhill reviewed (commented):
Thanks @jianzs!

I think we also need to add `device_control_env_var = "CUDA_VISIBLE_DEVICES"` to `CudaPlatformBase` and `RocmPlatform` right?

jianzs commented:
> Thanks @jianzs!
> 
> I think we also need to add `device_control_env_var = "CUDA_VISIBLE_DEVICES"` to `CudaPlatformBase` and `RocmPlatform` right?

Thanks for the suggestion, but both `CudaPlatformBase` and `RocmPlatform` already have this attribute defined.

https://github.com/vllm-project/vllm/blob/950b71186f9696a60f1a53cd1a033f7fef952500/vllm/platforms/cuda.py#L68-L74

https://github.com/vllm-project/vllm/blob/950b71186f9696a60f1a53cd1a033f7fef952500/vllm/platforms/rocm.py#L135-L142

jianzs commented:
@njhill  Any further questions?

njhill commented:
Ah :facepalm: sorry @jianzs, I hadn't noticed that!

njhill reviewed (commented):
@jianzs sorry for the delay getting back to this. Could we move `device_id_to_physical_device_id` to be a class method on `Platform`? Otherwise the change looks good to me.

wangxiyuan commented:
+1, move `device_id_to_physical_device_id` to `Platform` look more reasonable,

jianzs commented:
@njhill @wangxiyuan  Thank you for the suggestions. I've implemented the requested changes. Could you please review the updates?

jianzs committed:
feat: Use platform-agnostic device control for DP engine core

Refactor the DP engine core to use the platform-specific `device_control_env_var` attribute instead of hardcoding CUDA_VISIBLE_DEVICES. This change improves platform compatibility and code maintainability.

The update includes:
1. Moving the `device_id_to_physical_device_id` function to a shared utils file
2. Updating imports in CUDA and ROCm platform files
3. Replacing CUDA-specific environment variable setting with a platform-agnostic approach in the DPEngineCoreProc class

This refactoring enhances the flexibility of the codebase to support different platforms more seamlessly.

Signed-off-by: Jade Zheng <zheng.shoujian@outlook.com>

jianzs committed:
chore: lint code

Signed-off-by: Jade Zheng <zheng.shoujian@outlook.com>

njhill reviewed (approved):
Thanks @jianzs!
